compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -381,20 +392,47 @@ generateSrcFromLf thisPkgId pkgMap m = noLoc mod
     convType =
         \case
             LF.TVar tyVarName ->
-                HsTyVar NoExt NotPromoted $
-                mkRdrName $ LF.unTypeVarName tyVarName
-            LF.TCon LF.Qualified {..} ->
-                HsTyVar NoExt NotPromoted $
-                noLoc $
-                mkOrig
-                    (mkModule
-                         (getUnitId qualPackage)
-                         (mkModuleName $
-                          T.unpack $ LF.moduleNameString qualModule))
-                    (mkOccName varName $
-                     T.unpack $ T.intercalate "." $ LF.unTypeConName qualObject)
+                HsTyVar NoExt NotPromoted $ mkRdrName $ LF.unTypeVarName tyVarName
+            LF.TCon LF.Qualified {..}
+                | [name] <- LF.unTypeConName qualObject ->
+                    HsTyVar NoExt NotPromoted $
+                    noLoc $
+                    mkOrig
+                        (mkModule
+                             (getUnitId qualPackage)
+                             (mkModuleName $
+                              T.unpack $ LF.moduleNameString qualModule))
+                        (mkOccName varName $ T.unpack name)
+                | n@[_name0, _name1] <- LF.unTypeConName qualObject ->
+                    let fs =
+                            fromMaybe
+                                (error $
+                                 "Internal error: Could not find generated record type: " <>
+                                 (T.unpack $ T.intercalate "." n)) $
+                            MS.lookup n sumProdRecords
+                     in HsRecTy
+                            NoExt
+                            [ noLoc $
+                            ConDeclField
+                                { cd_fld_ext = noExt
+                                , cd_fld_names =
+                                      [ noLoc $
+                                        FieldOcc
+                                            { extFieldOcc = noExt
+                                            , rdrNameFieldOcc =
+                                                  mkRdrName $
+                                                  LF.unFieldName fieldName
+                                            }
+                                      ]
+                                , cd_fld_type = noLoc $ convType fieldTy
+                                , cd_fld_doc = Nothing
+                                }
+                            | (fieldName, fieldTy) <- fs
+                            ]
+                | cs <- LF.unTypeConName qualObject -> errTooManyNameComponents cs
             LF.TApp ty1 ty2 ->
-                HsAppTy NoExt (noLoc $ convType ty1) (noLoc $ convType ty2)
+                HsParTy noExt $
```

Comment:
What's the difference between `NoExt` and `noExt`?

Reply:
none I think.